keylime/cert_utils.py

```diff
@@ -1,13 +1,133 @@
+
+
+def x509_pem_cert(pem_cert_data):
+    if isinstance(pem_cert_data, (bytes, bytearray)):
+        pem_cert_data = pem_cert_data.decode("utf-8")
+    try:
+        cert = x509.load_pem_x509_certificate(data=pem_cert_data, backend=default_backend())
+        return cert
+    except Exception:
+        return x509_der_cert(pem.readPemFromFile(io.StringIO(pem_cert_data)))
+
+
+def verify_ek(ekcert, tpm_cert_store=config.get("tenant", "tpm_cert_store")):
+    """Verify that the provided EK certificate is signed by a trusted root
+    :param ekcert: The Endorsement Key certificate in DER format
+    :returns: True if the certificate can be verified, false otherwise
+    """
+    try:
+        ek509 = x509_der_cert(ekcert)
+
+        trusted_certs = tpm_ek_ca.cert_loader(tpm_cert_store)
+        for cert_file, pem_cert in trusted_certs.items():
+            signcert = x509_pem_cert(pem_cert)
+            if ek509.issuer != signcert.subject:
+                continue
+
+            signcert_pubkey = signcert.public_key()
+            try:
+                if isinstance(signcert_pubkey, RSAPublicKey):
+                    signcert_pubkey.verify(
+                        ek509.signature,
+                        ek509.tbs_certificate_bytes,
+                        padding.PKCS1v15(),
+                        ek509.signature_hash_algorithm,
+                    )
+                elif isinstance(signcert_pubkey, EllipticCurvePublicKey):
+                    signcert_pubkey.verify(
+                        ek509.signature,
+                        ek509.tbs_certificate_bytes,
+                        ec.ECDSA(ek509.signature_hash_algorithm),
+                    )
+                else:
+                    logger.warning("Unsupported public key type: %s", type(signcert_pubkey))
+                    continue
+            except crypto_exceptions.InvalidSignature:
+                continue
+
+            logger.debug("EK cert matched cert: %s", cert_file)
+            return True
+    except Exception as e:
+        # Log the exception so we don't lose the raw message
+        logger.exception(e)
+        raise Exception("Error processing ek/ekcert. Does this TPM have a valid EK?").with_traceback(sys.exc_info()[2])
+
+    logger.error("No Root CA matched EK Certificate")
+    return False
+
+
+def verify_ek_script(script, env, cwd):
+    if script is None:
+        logger.warning("External check script (%s) not specified", script)
+        return False
+
+    script_path = os.path.abspath(script)
+    if not os.path.isfile(script_path):
+        if cwd is None or not os.path.isfile(os.path.abspath(os.path.join(cwd, script))):
+            logger.warning("External check script (%s) not found; please make sure its path is correct", script)
+            return False
+        script_path = os.path.abspath(os.path.join(cwd, script))
+
+    try:
+        proc = subprocess.run(
+            [script_path],
+            env=env,
+            shell=True,
```

Comment:
This is dangerous in general and even if the old code had that we can it with this refactor.
I think the script is never called with any options, so we don't need that.

Reply:
Done.